src/nupic/algorithms/ColumnPooler.cpp

```diff
@@ -0,0 +1,565 @@
+ * this software code, the following terms and conditions apply:
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU Affero Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero Public License
+ * along with this program.  If not, see http://www.gnu.org/licenses.
+ *
+ * http://numenta.org/licenses/
+ * ----------------------------------------------------------------------
+ */
+
+/** @file
+ * Implementation of ColumnPooler
+ */
+
+#include <iostream>
+#include <vector>
+#include <string>
+#include <algorithm>
+#include <iomanip> // std::setprecision
+#include <functional> // function
+
+// #include <nupic/algorithms/ColumnPooler.hpp>
+#include <nupic/types/Types.hpp>
+#include <nupic/types/Serializable.hpp>
+#include <nupic/types/Sdr.hpp>
+#include <nupic/utils/SdrMetrics.hpp>
+#include <nupic/algorithms/Connections.hpp>
+#include <nupic/algorithms/TemporalMemory.hpp>
+#include <nupic/math/Math.hpp>
+#include <nupic/math/Topology.hpp>
+
+namespace nupic {
+namespace algorithms {
+namespace column_pooler {
+
+using namespace std;
+using namespace nupic;
+using namespace nupic::math::topology;
+using namespace nupic::algorithms::connections;
+using nupic::algorithms::temporal_memory::TemporalMemory;
+
+
+// TODO: Connections learning rules are different.
+
+
+/* Topology( location, potentialPool ) */
+typedef function<void(SDR&, SDR&)> Topology_t;
+
+class ColumnPooler // : public Serializable
+{
+private:
+  vector<UInt> proximalInputDimensions_;
+  vector<UInt> distalInputDimensions_;
+  vector<UInt> inhibitionDimensions_;
+  vector<UInt> cellDimensions_;
+  UInt         cellsPerInhibitionArea_;
+  UInt         proximalSegments_;
+
+  vector<UInt32> rawOverlaps_;
+  vector<UInt> proximalMaxSegment_;
+
+  Random rng_;
+  vector<Real> tieBreaker_;
+  UInt iterationNum_;
+  UInt iterationLearnNum_;
+
+public:
+  const vector<UInt> &proximalInputDimensions = proximalInputDimensions_;
+  const vector<UInt> &distalInputDimensions   = distalInputDimensions_;
+  const vector<UInt> &inhibitionDimensions    = inhibitionDimensions_;
+  const UInt         &cellsPerInhibitionArea  = cellsPerInhibitionArea_;
+  const vector<UInt> &cellDimensions          = cellDimensions_;
+  const UInt         &proximalSegments        = proximalSegments_;
+
+  Real sparsity;
+
+  UInt       proximalSegmentThreshold;
+  Permanence proximalIncrement;
+  Permanence proximalDecrement;
+  Permanence proximalSynapseThreshold;
+
+  SDR_ActivationFrequency *AF;
+  Real stability_rate;
+  Real fatigue_rate;
+  vector<Real> X_act;
+  vector<Real> X_inact;
```

Comment:
ditto, + more descriptive name please

Reply:
I don't know what to call these so that their name is descriptive...  Any suggestions?

These variables hold the state of the exponential moving averages for stability & fatigue.

Reply:
can you use MovingAverage, or SlidingWindow to do that? 
(activeCellsMovingAverage, ..?)

Reply:
I dont think so. These variables do an exponential moving average.